### PR TITLE
[TG-634] Logical expression building utility

### PR DESCRIPTION
**Pull request by LAJW**
This PR implements an `exprt` wrapper for constructing logical expression hierarchies using C++ operators.

Additionally, this PR contains small refactor on one of the most used string functions to show that it's working as intended.

For example, it lets following code:
```c++
exprt res_length=plus_exprt( 
  s1.length(), minus_exprt(end_index, start_index));
binary_relation_exprt prem(end_index, ID_gt, start_index); 
implies_exprt a1(prem, equal_exprt(res.length(), res_length));
```

be expressed this way:

```c++
exprt a1 = end_index > start_index >>=
           axt(s1.length()) + end_index - start_index == axt(res.length());

```

Part 2 of [String Refinement overhaul PR](https://github.com/diffblue/cbmc/pull/1399/commits)

### CODE REVIEW

**Comment by LAJW** [comment]
@reuk @jasigal @romainbrenguier Take a look if you have some spare time on your hands.

**Review by reuk — APPROVED** [review]
Nice and idiomatic :+1:

File: src/solvers/refinement/axt.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    :expr_(expr) { }
+  axt operator>(const axt& rhs) const
+  { return axt(binary_relation_exprt(expr_, ID_gt, rhs.expr_)); }
+  axt operator<(const axt& rhs) const
+  { return axt(binary_relation_exprt(expr_, ID_lt, rhs.expr_)); }
+  axt operator>=(const axt& rhs) const
+  { return !(*this<rhs); }
+  axt operator<=(const axt& rhs) const
+  { return !(*this>rhs); }
+  axt operator==(const axt& rhs) const
```
Comment (romainbrenguier) [inline src/solvers/refinement/axt.h]: could we also add `!=` for `notequal_exprt`?

File: src/solvers/refinement/axt.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  axt operator+(const axt& rhs) const
+  { return axt(plus_exprt(expr_, rhs.expr_)); }
+  axt operator-(const axt& rhs) const
+  { return axt(minus_exprt(expr_, rhs.expr_)); }
+  axt operator&&(const axt& rhs) const
+  { return axt(and_exprt(expr_, rhs.expr_)); }
+  axt operator||(const axt& rhs) const
+  { return axt(or_exprt(expr_, rhs.expr_)); }
+  axt operator[](const axt& index) const
+  { return axt(to_string_expr(expr_)[index.expr_]); }
```
Comment (romainbrenguier) [inline src/solvers/refinement/axt.h]: We may want to use this for `index_exprt` in case it's not a `string_exprt`

File: src/solvers/refinement/axt.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  axt operator<(const axt& rhs) const
+  { return axt(binary_relation_exprt(expr_, ID_lt, rhs.expr_)); }
+  axt operator>=(const axt& rhs) const
+  { return !(*this<rhs); }
+  axt operator<=(const axt& rhs) const
+  { return !(*this>rhs); }
+  axt operator==(const axt& rhs) const
+    { return axt(equal_exprt(expr_, rhs.expr_)); }
+  // Implication
+  axt operator>>=(const axt& rhs) const
```
Comment (jasigal) [inline src/solvers/refinement/axt.h]: Can we not use `->`?

File: src/solvers/refinement/axt.h (hunk 2 above)
Comment (jasigal) [inline src/solvers/refinement/axt.h]: Why do we cast to a string expression? Is this only going to be for the string solver?

**Review by jasigal — COMMENTED** [review]
Looks good :)

File: src/solvers/refinement/axt.h (hunk 2 above)
Comment (romainbrenguier) [reply src/solvers/refinement/axt.h]: For now it's only used in the string solver, but ultimately this should represent `index_of_exprt` (taking an `array_exprt` on the left) this will be unified once the new version of the string solver is merged as strings will be represented by arrays.

File: src/solvers/refinement/axt.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+{
+public:
+  axt(const exprt& expr) // NOLINT Use implicit conversions for best results
+    :expr_(expr) { }
+  axt operator>(const axt& rhs) const
+  { return axt(binary_relation_exprt(expr_, ID_gt, rhs.expr_)); }
+  axt operator<(const axt& rhs) const
+  { return axt(binary_relation_exprt(expr_, ID_lt, rhs.expr_)); }
+  axt operator>=(const axt& rhs) const
+  { return !(*this<rhs); }
```
Comment (jasigal) [inline src/solvers/refinement/axt.h]: Should use `axt(binary_relation_exprt(expr_, ID_ge, rhs.expr_))`

File: src/solvers/refinement/axt.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  axt operator<=(const axt& rhs) const
+  { return !(*this>rhs); }
```
Comment (jasigal) [inline src/solvers/refinement/axt.h]: Should use `axt(binary_relation_exprt(expr_, ID_le, rhs.expr_))`

**Review by jasigal — CHANGES_REQUESTED** [review]

File: src/solvers/refinement/axt.h (hunk 3 above)
Comment (LAJW) [reply src/solvers/refinement/axt.h]: No, because `->` is an unary operator. An alias for `(*ptr).<method_name>`

**Comment by martin-cs** [comment]
We had this code; pretty much exactly.  It was called oexpr and in util/oexpr.h.  It was deleted by PR #347 : https://github.com/diffblue/cbmc/commit/2db625 as part of a clean up in moving to a new C++ version, having been marked as "old" https://github.com/diffblue/cbmc/commit/346495 .

As I recall, I thought of it and suggested it to Daniel on the basis you could have one piece of code that was both concrete or symbolic depending on the types used (of course you can also achieve this by doing everything symbolically and then constant folding via simply_expr, anyway, I thought I was clever at the time), he implemented it because, why not, it then sat there and wasn't used because it turns out to not be as useful as you'd originally think and then was removed because it wasn't being used.

I guess there are multiple conclusions one could draw from this story and I'll let you pick your own.

File: src/solvers/refinement/axt.h (hunk 2 above)
Comment (LAJW) [reply src/solvers/refinement/axt.h]: I shall wait then until they become one and the same thing.

File: src/solvers/refinement/axt.h [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+ Author: Diffblue Limited. All rights reserved.
+
+\*******************************************************************/
+#ifndef CPROVER_SOLVERS_REFINEMENT_AXT_H
+#define CPROVER_SOLVERS_REFINEMENT_AXT_H
+
+#include <util/expr.h>
+#include <util/string_expr.h>
+
```
Comment (martin-cs) [inline src/solvers/refinement/axt.h:13]: I would have thought this would be happier in util/ as nothing it does is specific to refinement.  Also, maybe we should just resurrect oexpr.

File: src/solvers/refinement/string_constraint_generator_concat.cpp [hunk 7]
```diff
@@ -12,6 +12,7 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 ///   strings
 
 #include <solvers/refinement/string_constraint_generator.h>
+#include "axt.h"
```
Comment (martin-cs) [inline src/solvers/refinement/string_constraint_generator_concat.cpp:15]: Please include in the usual style.

File: src/solvers/refinement/string_constraint_generator_concat.cpp [hunk 8]
```diff
@@ -39,32 +40,38 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
...
   const array_string_exprt &res,
-  const array_string_exprt &s1,
-  const array_string_exprt &s2,
-  const exprt &start_index,
-  const exprt &end_index)
+  const array_string_exprt &s1_,
+  const array_string_exprt &s2_,
+  const exprt &start_index_,
+  const exprt &end_index_)
 {
```
Comment (martin-cs) [inline src/solvers/refinement/string_constraint_generator_concat.cpp:47]: Why are these renamed?

**Review by martin-cs — CHANGES_REQUESTED** [review]
Over all ... I'm not convinced it gains us that much.  A slight reduction in characters in long lines but that is traded against obfuscating what is symbolic and what is literal.

As I said before, I used to think this was a great idea but I've not yet seen a compelling use-case.

**Comment by LAJW** [comment]
Change too controversial, was attempted before and rejected. Closing.
